### PR TITLE
[RPA-1307] BPMN error handling

### PR DESCRIPTION
**Pull request by seul88**
## Description
**Demo process using error messages content to handle its flow is attached as .rbex file in JIRA ticket's details.**

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts [hunk 1]
```diff
@@ -459,6 +459,7 @@ export class RuntimeService implements OnApplicationBootstrap, OnModuleDestroy {
                 ...request.variables,
                 tempFolder: this.getTempDirPath(),
                 userEmail: triggerData && 'userEmail' in triggerData ? triggerData.userEmail : '',
+                errorMessages: []
```
Comment (preludium) [inline runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts]: imo you don't need to initialize it at process startup

File: runbotics/runbotics-desktop/src/core/websocket/bpmn/runtime-subscriptions.service.ts [hunk 2]
```diff
@@ -55,6 +55,13 @@ export class RuntimeSubscriptionsService {
                     ).behaviour;
                     switch (event.activity.content.type) {
                         case BpmnElementType.ERROR_EVENT_DEFINITION:
+                            const variables = {};
+                            const prevMessages = event.activity.environment.variables['errorMessages'] as string[];
+                            prevMessages.push(desktopTask.output.description);
+                            variables['errorMessages'] = prevMessages;
+                            const setVariables = { ...event.activity.environment.variables, ...variables};
+                            event.activity.environment.assignVariables(setVariables);
+                            processInstanceEvent.error = desktopTask.output.description;
```
Comment (preludium) [inline runbotics/runbotics-desktop/src/core/websocket/bpmn/runtime-subscriptions.service.ts]: use this function for it imo https://github.com/runbotics/runbotics/blob/99a3c68536872a7009c2f5918856775fa2be848a/runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts#L642

File: runbotics/runbotics-orchestrator-ui/src/src-app/hooks/useOptions.ts [hunk 3]
```diff
@@ -52,7 +52,8 @@ const UTILS = [
 
 const VARIABLES = [
     'tempFolder',
-    'userEmail'
+    'userEmail',
+    'errorMessages'
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/hooks/useOptions.ts]: imo rename it to `caughtErrorMessage` we will store just the most recent error message

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-desktop/src/action/variable/variable.action-handler.ts [hunk 4]
```diff
@@ -38,11 +38,7 @@ export default class VariableActionHandler extends StatelessActionHandler {
     ): Promise<AssignVariableActionOutput> {
         const variables = {};
         variables[input.variable] = input.value;
-        // assign to local scope
-        const setVariables = { ...request.executionContext.environment.variables, ...variables };
-        request.executionContext.environment.assignVariables(setVariables);
 
-        // assign to global scope
```
Comment (preludium) [inline runbotics/runbotics-desktop/src/action/variable/variable.action-handler.ts:45]: im not sure about it. debug what the difference between so called local and global scope

**Review by star-in-the-jar — APPROVED** [review]

**Pull request merged**